src/plexil-adapter/OwInterface.cpp

```diff
@@ -746,140 +646,235 @@ void OwInterface::takePicture (int id)
 {
   if (! mark_operation_running (Op_TakePicture, id)) return;
   std_msgs::Empty msg;
-  ROS_INFO ("  Capturing stereo image using left image trigger.");
+  ROS_INFO ("Capturing stereo image using left image trigger.");
   thread fault_thread (monitor_for_faults, Op_TakePicture);
   fault_thread.detach();
   m_leftImageTriggerPublisher->publish (msg);
 }
 
+void OwInterface::deliver (double x, double y, double z, int id)
+{
+  if (! mark_operation_running (Op_Deliver, id)) return;
+  thread action_thread (&OwInterface::deliverAction, this, x, y, z, id);
+  action_thread.detach();
+}
+
+template <int OpIndex, class ActionClient, class Goal,
+          class ResultPtr, class FeedbackPtr>
+void OwInterface::runAction (const string& opname,
+                             std::unique_ptr<ActionClient>& ac,
+                             const Goal& goal, int id)
+{
+  thread fault_thread (monitor_for_faults, opname);
+  if (ac) {
+    ac->sendGoal (goal,
+                  action_done_cb<OpIndex, ResultPtr>,
+                  active_cb<OpIndex>,
+                  action_feedback_cb<FeedbackPtr>);
+  }
+  else {
+    ROS_ERROR ("%s action client was null!", opname.c_str());
+    return;
+  }
+
+  // Wait for the action to return
+  bool finished_before_timeout =
+    ac->waitForResult (ros::Duration (ActionTimeoutSecs));
```

Comment:
This is a serious issue: every action will block autonomy for this timeout regardless of how fast it finishes.

Reply:
Just use 0 for the duration, this should resolve it: 
```suggestion
    ac->waitForResult (ros::Duration (0));
```

Reply:
Will test with this...

Reply:
It worked fine, so I made this change.  Now, autonomy will wait indefinitely for actions to finish.  This is not good long-term, but the proper solution is out of scope for this ticket, and Release 8.